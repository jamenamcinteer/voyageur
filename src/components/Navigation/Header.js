import React, { useState } from "react";
import { Link } from "react-router-dom";
import styled from "styled-components";
import theme from "../../theme";
import PropTypes from "prop-types";

const HeaderBackground = styled.div`
  background-color: ${props => props.theme.headerBackground};
  padding: 10px 20px;
  // text-align: center;
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const HeaderText = styled.h1`
  font-size: 1.2em;
  font-weight: normal;
  color: ${props => props.theme.headerForeground};
`;

const ProfileButton = styled.button`
  background: 0;
  border: 0;
  padding: 0;
  border: 3px solid
    ${props =>
      props.selected
        ? props.theme.themeColor
        : props.theme.themeColorSecondary};
  border-radius: 100px;

  &:focus {
    outline: 0;
    border: 3px solid ${props => props.theme.focusBorder};
  }
`;
const ProfileImage = styled.img`
  border-radius: 100px;
  width: 45px;
  height: 45px;
  display: block;
`;
const ProfileMenu = styled.div`
  position: absolute;
  background: #fff;
  z-index: 1;
  padding: 0;
  right: 0;
  top: 68px;
  width: calc(100vw - 50px);
  height: calc(100vh - 69px);
  box-shadow: -3px 3px 4px 2px rgba(0, 0, 0, 0.2);
  max-width: 400px;
`;
const ProfileMenuLink = styled.a`
  color: ${props => props.theme.darkFont};
  text-decoration: none;
  border-bottom: 1px solid ${props => props.theme.themeColorSecondary};
  display: block;
  padding: 20px;
  text-align: left;
  font-size: 1.2em;
`;

const Header = props => {
  const [menuOpen, setMenuOpen] = useState(false);

  const svgStyles = {
    fill: theme.headerForeground,
    width: "20px",
    height: "20px",
    float: "left",
    // marginTop: "15px"
    marginTop: "2px"
  };

  return (
    <HeaderBackground>
      {!props.backHide && (
        <Link to={props.backTo}>
          <svg
            version="1.1"
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 199.404 199.404"
            style={svgStyles}
            aria-label="Go Back"
          >
            <g>
              <polygon points="135.412,0 35.709,99.702 135.412,199.404 163.695,171.119 92.277,99.702 163.695,28.285 	" />
            </g>
          </svg>
        </Link>
      )}
      <HeaderText>{props.title}</HeaderText>
      <ProfileButton onClick={e => setMenuOpen(!menuOpen)} selected={menuOpen}>
        <ProfileImage src={props.auth.photoURL} alt="Account Options" />
        {menuOpen && (
          <ProfileMenu>
            <ProfileMenuLink href="/auth/logout">Sign Out</ProfileMenuLink>
          </ProfileMenu>
        )}
      </ProfileButton>
    </HeaderBackground>
  );
};

Header.propTypes = {
  title: PropTypes.string.isRequired,
  auth: PropTypes.object.isRequired,
  backHide: PropTypes.bool,
  backTo: PropTypes.string
};

export default Header;
