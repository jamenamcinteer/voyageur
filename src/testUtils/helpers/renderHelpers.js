import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import theme from "../../theme";
import "jest-styled-components";
import { createStore } from "redux";
import { Provider } from "react-redux";
import tripsReducer, { tripsReducerDefaultState } from "../../reducers/trips";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";

export function renderWithTheme(component) {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
}

export function renderWithReduxRouterAndTheme(
  ui,
  {
    initialState,
    store = createStore(tripsReducer, tripsReducerDefaultState),
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) {
  return {
    ...render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <Router history={history}>{ui}</Router>
        </ThemeProvider>
      </Provider>
    ),
    store,
    history
  };
}
