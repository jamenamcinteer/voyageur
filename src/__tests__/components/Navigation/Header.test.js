import React from "react";
import { cleanup } from "@testing-library/react";
import Header from "../../../components/Navigation/Header";
import { renderWithReduxRouterAndTheme } from "../../../testUtils/helpers/renderHelpers";
import { toBeInTheDocument } from "@testing-library/jest-dom";

expect.extend({ toBeInTheDocument });

afterEach(cleanup);

test("should render with theme and passed props and with back button", () => {
  const auth = {
    photoURL: "/some-image.png"
  };
  const { asFragment } = renderWithReduxRouterAndTheme(
    <Header title="Test Header" auth={auth} backTo="/" />
  );
  expect(asFragment()).toMatchSnapshot();
});

test("should render with theme and passed props and with no back button", () => {
  const auth = {
    photoURL: "/some-image.png"
  };
  const { asFragment } = renderWithReduxRouterAndTheme(
    <Header title="Test Header" auth={auth} backHide={true} />
  );
  expect(asFragment()).toMatchSnapshot();
});

test("should open and close profile menu correctly", () => {
  const auth = {
    photoURL: "/some-image.png"
  };
  const {
    getByAltText,
    getByText,
    queryByText
  } = renderWithReduxRouterAndTheme(
    <Header title="Test Header" auth={auth} backTo="/" />
  );

  getByAltText("Account Options").click();
  expect(getByText("Sign Out")).toBeInTheDocument();

  getByAltText("Account Options").click();
  expect(queryByText("Sign Out")).toBeNull();
});
