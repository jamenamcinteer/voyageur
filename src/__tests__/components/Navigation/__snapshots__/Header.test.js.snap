// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with theme and passed props and with back button 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #dfe7ed;
  padding: 10px 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  font-size: 1.2em;
  font-weight: normal;
  color: #4b5c6b;
}

.c2 {
  background: 0;
  border: 0;
  padding: 0;
  border: 3px solid #c7d2db;
  border-radius: 100px;
}

.c2:focus {
  outline: 0;
  border: 3px solid #A63E26;
}

.c3 {
  border-radius: 100px;
  width: 45px;
  height: 45px;
  display: block;
}

<div
    class="c0"
  >
    <a
      href="/"
    >
      <svg
        aria-label="Go Back"
        style="fill: #4b5c6b; width: 20px; height: 20px; float: left; margin-top: 2px;"
        version="1.1"
        viewBox="0 0 199.404 199.404"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <polygon
            points="135.412,0 35.709,99.702 135.412,199.404 163.695,171.119 92.277,99.702 163.695,28.285 	"
          />
        </g>
      </svg>
    </a>
    <h1
      class="c1"
    >
      Test Header
    </h1>
    <button
      class="c2"
    >
      <img
        alt="Account Options"
        class="c3"
        src="/some-image.png"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`should render with theme and passed props and with no back button 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #dfe7ed;
  padding: 10px 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  font-size: 1.2em;
  font-weight: normal;
  color: #4b5c6b;
}

.c2 {
  background: 0;
  border: 0;
  padding: 0;
  border: 3px solid #c7d2db;
  border-radius: 100px;
}

.c2:focus {
  outline: 0;
  border: 3px solid #A63E26;
}

.c3 {
  border-radius: 100px;
  width: 45px;
  height: 45px;
  display: block;
}

<div
    class="c0"
  >
    <h1
      class="c1"
    >
      Test Header
    </h1>
    <button
      class="c2"
    >
      <img
        alt="Account Options"
        class="c3"
        src="/some-image.png"
      />
    </button>
  </div>
</DocumentFragment>
`;
